### PR TITLE
feat: Add format_engine to Representable::Object

### PR DESCRIPTION
**Pull request by dmuneras**
## Description

Hi,

I was trying to use `Representable::Object` to transform an object using another one as a source. But, when I was trying to create the represented it started to throw an error:

Screenshot 2021-08-04 at 15 21 55

But after adding this patch I can get it working:

Screenshot 2021-08-04 at 15 27 43

In case this is not a proper solution for the issue, I will appreciate it if you can point me in the right direction to achieve what I want which is decorating the object and returning a new object with only the info I know for it purpose. 

Best regards,
Daniel.

### CODE REVIEW

**Comment by yogeshjain999** [comment]
Thanks for the PR. 🍻 could also add a test for it ?

It seems `ObjectTest` doesn't have an example defining a representer with `Representable::Decorator`, it could have caught this error.

**Comment by dmuneras** [comment]
> Thanks for the PR. 🍻 could also add a test for it ?
> 
> It seems `ObjectTest` doesn't have an example defining a representer with `Representable::Decorator`, it could have caught this error.

Sure, I will do it later today.

**Comment by dmuneras** [comment]
> Thanks for the PR. 🍻 could also add a test for it?
> 
> It seems `ObjectTest` doesn't have an example defining a representer with `Representable::Decorator`, it could have caught this error.

@yogeshjain999  sorry for the delay. I hope the two tests I added are ok, if not let me know and I will update them (in less than 14 days this time 😅 )

Best regards,
Daniel.

File: test/object_test.rb [hunk 1]
```diff
@@ -57,4 +58,19 @@ class ObjectTest < MiniTest::Spec
       _(source.album.songs[0].title).must_equal 1
     end
   end
-end
\ No newline at end of file
+
+  describe ".format_engine" do
+    let(:decorator) do
```
Comment (yogeshjain999) [inline test/object_test.rb:63]: @dmuneras You can use existing `representer!` helper to build the representer.

```ruby
representer!(module: Representable::Object, decorator: true) do
  property :title
end

it { representer.prepare(Song.new).to_object }
```

File: test/object_test.rb [hunk 2]
```diff
@@ -57,4 +58,19 @@ class ObjectTest < MiniTest::Spec
...
+        include Representable::Object
+
+        property :city, as: :city_name
+      end
+
+      DecoratedObject
+    end
+
+    it { decorator.must_equal(ObjectTest::DecoratedObject) }
+    it { decorator.format_engine.must_equal(Representable::Object) }
```
Comment (yogeshjain999) [inline test/object_test.rb:74]: @dmuneras We should verify if `representer` renders correctly using `to_object`. You can check out the other test cases in this file for example.

Also, I don't think we need to test `.format_engine` as it's an internal API (as long as the `representer` generates correct result).

File: test/object_test.rb [hunk 3]
```diff
@@ -57,4 +58,19 @@ class ObjectTest < MiniTest::Spec
       _(source.album.songs[0].title).must_equal 1
     end
   end
-end
\ No newline at end of file
+
+  describe ".format_engine" do
```
Comment (yogeshjain999) [inline test/object_test.rb:62]: @dmuneras This is more like a test for using `Representable::Object` within decorator (as per the issue you've encountered). You can change the description here to mention that in order to not forget it later 😄 

For example,

```ruby
describe "renders object with decorator" do
end
```

**Comment by dmuneras** [comment]
I just read all the comments, gonna work on it later today. Thanks. 😄

**Comment by dmuneras** [comment]
Hi @yogeshjain999! 

I updated the tests according to what I understood was required. I have a doubt about this line (https://github.com/trailblazer/representable/pull/252/files#diff-20931e08025985d29965f2979d8b0fcba629544ceb0df4246f7769eb1daac0bbR69). It works but I don't know if it is the correct approach.

File: test/object_test.rb [hunk 4]
```diff
@@ -51,10 +57,27 @@ class ObjectTest < MiniTest::Spec
...
+
+      _(source.album.name).must_equal("Live")
+      _(source.album.songs[0].title).must_equal(1)
+    end
+
+    describe "when used as decorator" do
+      representer!(name: :decorator ,module: Representable::Object, decorator: true) do
+        property :name, getter: ->(represented:, **) {
+          represented.name = "#{represented.name} is the best team of America"
+        }
```
Comment (apotonick) [inline test/object_test.rb:71]: Why is the `getter` setting a value?

Comment (dmuneras) [reply test/object_test.rb]: Hi @apotonick,

I just changed the test to use the same logic in the existing test, but using `Representable::Object` as a decorator. 

As an honest answer to your question: I just pushed that commit and add the comment to be able to get some help to understand what is the goal of `to_object`. It looked wrong to set values from a lambda passed in an option called `getter`. But I didn't know which option makes sense when using a decorator to call `to_object`

I created this PR to be able to use from `from_object` with allows me to instantiate an object base on a source one. 

```
[3] pry(#<#to_object::when used as decorator>)> source = Album.new("Atletico Nacional")
=> #<struct ObjectTest::Album name="Atletico Nacional", songs=nil>
[4] pry(#<#to_object::when used as decorator>)> new_object = decorator.prepare(Album.new).from_object(source)
=> #<struct ObjectTest::Album name="Atletico Nacional", songs=nil>
[5] pry(#<#to_object::when used as decorator>)> new_object.object_id
=> 1760
[6] pry(#<#to_object::when used as decorator>)> source.object_id
=> 1780
[7] pry(#<#to_object::when used as decorator>)>
``` 

But, according to what I understand, `to_object` does modify the passed object.

**Comment by dmuneras** [comment]
Thanks for the patience @yogeshjain999. I will wait for this to be merged and released to remove my patch 😄

**Comment by dmuneras** [comment]
Hi @yogeshjain999,

Is there anything blocking this one from getting merged? just checking if I can do something to get it merged.

Best regards,
Daniel.

**Comment by yogeshjain999** [comment]
Ah, I almost forgot about this @dmuneras 
Sorry for the delay, but want to know @apotonick's thoughts before merging this.

**Comment by dmuneras** [comment]
Hi @yogeshjain999! - I think this won't move forward, so I will close it just to clean up. Thanks a lot for your feedback and have a good weekend.

**Pull request closed without merging**